### PR TITLE
Changes/nyholku

### PR DESCRIPTION
**Pull request by nyholku**
Minimal changes required to be able to write the OCC Bottle example in Java.
I had to rework all commits (a lot of work) hope it does not create too much work for you. 
I also trimmed things down a bit so that this is hopefully a minimal set of functionality  at this stage so almost everything is in use in the Bottle example and thus tested to some extent.
Note that this is a completely new branch, I will delete all the topic branches from repo as soon as you merge these, I don't think the history of my attempts and mixups is interesting.
For the next changes I will create a new branch based on your then latest master.

### CODE REVIEW

File: occjava/src/gp.i [hunk 1]
```diff
@@ -380,5 +496,57 @@ class gp_Parab
...
+%typemap(javaout) gp_Pnt, const gp_Pnt&
+{
+	return $jnicall;
+}
+
+
+%rename(GP) gp;
+%typemap(javacode) gp
+%{	
+	final public static double[] Origin = {0, 0, 0};
```
Comment (jeromerobert) [inline occjava/src/gp.i]: Should be ORIGIN to match Java Coding Conventions

File: occjava/src/BRepBuilderAPI.i [hunk 2]
```diff
@@ -220,3 +222,12 @@ class BRepBuilderAPI_NurbsConvert : public BRepBuilderAPI_ModifyShape
 		const Standard_Boolean Copy = Standard_False) ;
 };
 
+class BRepOffsetAPI_ThruSections  : public BRepBuilderAPI_MakeShape
+{
+public:
+	BRepOffsetAPI_ThruSections(const Standard_Boolean isSolid = Standard_False, const Standard_Boolean ruled = Standard_False, const Standard_Real pres3d = 1.0e-06);
+	void AddWire (const TopoDS_Wire& wire) ;
```
Comment (jeromerobert) [inline occjava/src/BRepBuilderAPI.i]: AddWire and CheckCompatibility should start with a lower case letter to match java coding conventions.

File: occjava/src/BRepLib.i [hunk 3]
```diff
@@ -27,5 +27,7 @@ class BRepLib
     public:
     %rename(encodeRegularity) EncodeRegularity;
     static void EncodeRegularity(const TopoDS_Shape& S,const Standard_Real TolAng = 1.0e-10) ;
+    static Standard_Boolean BuildCurves3d (const TopoDS_Shape& S) ;
```
Comment (jeromerobert) [inline occjava/src/BRepLib.i]: Should start with lower case

File: occjava/src/BRepOffsetAPI.i [hunk 4]
```diff
@@ -54,3 +56,20 @@ class BRepOffsetAPI_NormalProjection: public BRepBuilderAPI_MakeShape
...
+};
+
+class BRepOffsetAPI_MakeThickSolid  : public BRepOffsetAPI_MakeOffsetShape {
+public:
+	BRepOffsetAPI_MakeThickSolid(
+		const TopoDS_Shape& S, 
+		const TopTools_ListOfShape& ClosingFaces, 
+		const Standard_Real Offset, 
+		const Standard_Real Tol, 
+		const BRepOffset_Mode Mode = BRepOffset_Skin,
```
Comment (jeromerobert) [inline occjava/src/BRepOffsetAPI.i]: BRepOffset_Mode is not wrapped so I guess you don't need it. So all optional argument should be removed here. Wrapping BRepOffset_Mode and adding other argument could be done later.

File: occjava/src/GC.i [hunk 5]
```diff
@@ -16,11 +16,34 @@
...
+ 
+class  GCE2d_Root {
+ 	 public:
+ };
+
+class GCE2d_MakeSegment  : public GCE2d_Root
+{
+	public:
+  	GCE2d_MakeSegment(const gp_Pnt2d& P1, const gp_Pnt2d& P2);
+	Handle_Geom2d_TrimmedCurve& Value () const;
```
Comment (jeromerobert) [inline occjava/src/GC.i]: Shoud be renamed to `value` to match java coding conventions

File: occjava/src/GC.i [hunk 6]
```diff
@@ -16,11 +16,34 @@
...
+ };
+ 
+class GC_MakeSegment {
+    public:
+    GC_MakeSegment(const gp_Pnt& P1, const gp_Pnt& P2);
+ 	const Handle_Geom_TrimmedCurve& Value() const;
+ };
+ 
+class  GCE2d_Root {
+ 	 public:
```
Comment (jeromerobert) [inline occjava/src/GC.i]: It's not virtual pure in OCE/OCC but IMHO it should be. So here I would add a virtual pure constructor, so no java code is generated to allow creation of this object.

File: occjava/src/Geom.i [hunk 7]
```diff
@@ -293,3 +305,64 @@ class Handle_Geom_TrimmedCurve : public Handle_Geom_BoundedCurve {
 class Handle_Geom_Plane: public Handle_Geom_Surface {
        Handle_Geom_Plane()=0;
 };
+
+%extend Handle_Geom_Plane {
+	gp_Pnt Location()
```
Comment (jeromerobert) [inline occjava/src/Geom.i]: `Location` must be wrapped in `ElementarySurface` (as in OCC) and renamed to `location`.

File: occjava/src/Geom.i [hunk 8]
```diff
@@ -293,3 +305,64 @@ class Handle_Geom_TrimmedCurve : public Handle_Geom_BoundedCurve {
...
+};
+
+%extend Handle_Geom2d_Ellipse
+{
+	Handle_Geom2d_Ellipse(const gp_Ax22d& Axis, const Standard_Real MajorRadius, const Standard_Real MinorRadius) 
+	{
+	return new Handle_Geom2d_Ellipse(new Geom2d_Ellipse(Axis,MajorRadius,MinorRadius));
+	}
+	
+    gp_Pnt2d Value(const Standard_Real U) const
```
Comment (jeromerobert) [inline occjava/src/Geom.i]: Must be moved to `Geom2d_Curve` (as in OCC) and renamed to `value`

File: occjava/src/gp.i [hunk 9]
```diff
@@ -315,6 +430,7 @@ class gp_Trsf
 	gp_Trsf();
 	void SetRotation(const gp_Ax1& A1,const Standard_Real Ang) ;
 	void SetTranslation(const gp_Vec& V) ;
+	void SetMirror(const gp_Ax1& P);
```
Comment (jeromerobert) [inline occjava/src/gp.i]: To `%rename` to `setMirror`

**Comment by jeromerobert** [comment]
Just in case of you don't know it yet. It's possible to edit existing commit without pain, using the `git rebase -i` command and use the `edit` option for commit that need to be changed.
I can also do the changes for you if you prefer.

**Comment by nyholku** [comment]
Hi, 
thanks for the valuable and patient guidance. 

 I just did the changes but have not yet committed / pushed them.

So you prefer that I rework the commit and not add new commits that fix the issues you spotted/indicated?

 I can do that no problem.

**Comment by jeromerobert** [comment]
> So you prefer that I rework the commit and not add new commits that fix the issues you
> spotted/indicated?

Yes I would. But if you add one commit by issue, it's easy for me to squash them after, using the `git rebase -i` command, so you can do as you prefer.

**Comment by nyholku** [comment]
I will do the squashing so I will learn in the process. I will be later today I hope.

**Comment by nyholku** [comment]
I'm closing this cause ...  after three hours struggling with interactive rebase I got stuck with a mistaken squash that I could not split and wanting to present a clean history I just had to resort to a new branch and got it all done in 15 minutes. I've got a lot to learn with git I guess.

**Pull request closed without merging**